rosdoc2/verbs/build/builders/sphinx_builder.py

```diff
@@ -366,10 +409,55 @@ def build(self, *, doc_build_folder, output_staging_directory):
             if package_name != self.build_context.package.name
         ]
 
+        # If the package has build type `ament_python`, or if the user configured
+        # to run `sphinx-apidoc`, then invoke `sphinx-apidoc` before building
+        if (
+            self.build_context.build_type == 'ament_python' or
+            self.build_context.always_run_sphinx_apidoc
+        ):
+            package_xml_directory = os.path.dirname(self.build_context.package.filename)
+            # If 'python_source' is specified, construct 'package_src_directory' from it
+            if self.build_context.python_source is not None:
+                package_src_directory = \
+                    os.path.join(
+                        package_xml_directory,
+                        self.build_context.python_source)
+            # If not provided, try to find the package source direcotry
+            else:
+                package_list = setuptools.find_packages(where=package_xml_directory)
+                if self.build_context.package.name in package_list:
+                    package_src_directory = os.path.join(
+                        package_xml_directory,
+                        self.build_context.package.name)
+                else:
+                    package_src_directory = None
+
+            if not package_src_directory or not os.path.isdir(package_src_directory):
+                raise RuntimeError(
+                    "Could not locate source directory to invoke sphinx-apidoc in. "
+                    "If this is package does not have a standard Python package layout, "
+                    "please specify the Python source in 'rosdoc2.yaml'.")
+            cmd = [
+                'sphinx-apidoc',
+                '-o', os.path.relpath(sourcedir, start=doc_build_folder),
+                '-e',  # Document each module in its own page.
+                os.path.abspath(package_src_directory),
+            ]
+            logger.info(
+                f"Running sphinx-apidoc: '{' '.join(cmd)}' in '{doc_build_folder}'"
+            )
+            completed_process = subprocess.run(cmd, cwd=doc_build_folder)
```

Comment:
On first run, the variable doc_build_folder has been created with the correct path, but that path has not actually been created yet. The creation occurs later in generate_wrapping_rosdoc2_sphinx_project_into_directory. You either need to create the directory, or move generate_wrapping_rosdoc2_sphinx_project_into_directory to occur prior to this call.

I'm not really sure what the etiquette is in the ROS world around what we used to call "drive-by reviews" in the Mozilla world. Hopefully this is useful and not inappropriate. I have been trying to setup a mock ros2 package with all of the bells and whistles, including various rosdoc2 setups, so I am encountering these issues.

Reply:
> I'm not really sure what the etiquette is in the ROS world around what we used to call "drive-by reviews"

Very much appreciated!  This PR has been on our long-term list to test and get in, but we just haven't gotten to it yet.  We are hoping to soon, so pointing out these problems is very helpful, thanks.